Fix for test_backward_node_failure: Error Handling in RPC Agent

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36620 Fixed error Regex Parsing for Node Failure Tests
* **#35263 Fix for test_backward_node_failure: Error Handling in RPC Agent**

Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

osalpekar committed:
Error Handling in RPC Agent

Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

[ghstack-poisoned]

osalpekar committed:
Update on "Error Handling in RPC Agent"


Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

[ghstack-poisoned]

pritamdamania87 reviewed (changes requested):
Can we add a unit test for this case?

osalpekar commented:
> Can we add a unit test for this case?

@pritamdamania87 This (in conjunction with https://github.com/pytorch/pytorch/pull/35394) are addressing `test_backward_node_failure` and `test_backward_node_failure_python_udf` in `dist_autograd_test.py`, which are currently flaky. Should we add more tests, or do these 2 cover this functionality?

rohan-varma commented:
@osalpekar Is it possible to add something like the following test in `ProcessGroupAgentTest`?

Thread 1 (background thread): Waits for `agent.shutdown()` to be called from another thread, by spinning or CV. Then, when seeing this, kicks off an RPC that will fail and be retried n times. Returns after all these retries fails
Thread 2 (Main): Calls agent.shutdown, then joins thread 1.

osalpekar committed:
Update on "Fix for test_backward_node_failure: Error Handling in RPC Agent"


Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

[ghstack-poisoned]

mrshenli commented:
Will this fix #35099 as well?

rohan-varma commented:
@mrshenli This PR in conjunction with https://github.com/pytorch/pytorch/pull/35394 should help fix the issue.

osalpekar commented:
Added an internal test case. One worker calls `sendWithRetries` using the faulty agent, so the first send fails. It then immediately calls `rpcAgent.shutdown`, so the agent is not alive when the first retry is attempted. We set an appropriate error on the future returned by `sendWithRetries` and assert the exception message is as expected.

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -106,8 +106,20 @@ void RpcAgent::retryExpiredRpcs() {
+      } catch (std::exception& e) {
+        std::string errorMsg =
+            c10::str("send() failed while retrying RPC with error: ", e.what());
+        earliestRpc->originalFuture_->setError(errorMsg);
```
calling setError() with the lock held might lead to deadlocks.
(not sure if that's related to the observed issue)

osalpekar replied:
Thanks, just realized this as well. Not related to the test failure, but I'll push a fix

osalpekar committed:
Update on "Fix for test_backward_node_failure: Error Handling in RPC Agent"


Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

[ghstack-poisoned]

osalpekar commented:
> question: if we've dropped and re-acquired the lock, will the iterator still be valid in line 134 below?

It's possible that `rpcRetryMap` was modified when we dropped the lock. It might make more sense to store the futures and excpetion messages in the catch block, and later call `setError` on all those futures without the lock (similar to what we did for adding callbacks).

osalpekar committed:
Update on "Fix for test_backward_node_failure: Error Handling in RPC Agent"


Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -303,6 +303,10 @@ class TORCH_API RpcAgent {
       futures;
 
+  // storing futures and exception messages for non-retriable error-ed futures.
+  std::vector<std::pair<std::shared_ptr<FutureMessage>, std::string>>
```
why is this a member var, and not a local function var?

rohan-varma commented:
Can we get this PR landed or is there something currently blocking it? Would definitely help in resolving a bunch of longstanding flaky tests.

osalpekar committed:
Update on "Fix for test_backward_node_failure: Error Handling in RPC Agent"


Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

[ghstack-poisoned]

jjlilley commented:
I think we should get this landed, but wondering why errorFutures is a member var, per my comment a week ago?

osalpekar commented:
> Can we get this PR landed or is there something currently blocking it? Would definitely help in resolving a bunch of longstanding flaky tests.

This stack is ready, but these alone will not fix some of the flaky tests (`test_backward_node_failure` should be working fine after this, `test_backward_node_failure_python_udf` will still fail infrequently). We need to either `shutdown` or allow send/recv after a listenLoop exception in PGA in order for `test_backward_node_failure_python_udf` to pass regularly. @jjlilley and I discussed and I have a fix on the way.

osalpekar commented:
> I think we should get this landed, but wondering why errorFutures is a member var, per my comment a week ago?

hm looks like I didn't commit that change here - I'll update this right now

osalpekar committed:
Update on "Fix for test_backward_node_failure: Error Handling in RPC Agent"


Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -78,6 +78,10 @@ std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
 
 void RpcAgent::retryExpiredRpcs() {
+  // storing futures and exception messages for non-retriable error-ed futures.
+  std::vector<std::pair<std::shared_ptr<FutureMessage>, std::string>>
```
just checking: the idea is that we only want this cleared with the lock held? (per line 171)
(it's not clear that this would strictly be necessary?)

Does it make sense to bump the scope up to inside the while loop, so guaranteed not to leak these between runs?
(maybe it doesn't matter...)

osalpekar replied:
We should be able to clear the vector without the lock held now that its a local function var.

osalpekar committed:
Update on "Fix for test_backward_node_failure: Error Handling in RPC Agent"


Process Group Agent throws an exception if a send attempt is made after the agent is shutdown. With retries, we should catch this exception and mark the original future with an error.

Differential Revision: [D20611412](https://our.internmc.facebook.com/intern/diff/D20611412/)

[ghstack-poisoned]

jjlilley reviewed (approved):
Looks great, thanks!
Let's get this in after the tests look reasonable.
